Fix `ExecutionMode.WATCHER` behaviour with `DbtTaskGroup`

tatiana opened the pull request:
While validating `TaskGroup` in a session with an [end-user](https://us-35700.app.gong.io/call?id=1548801879239330851&email_type=call-ready-notification&xtid=45fp79m8frjifyiyqp0) with Cosmos [v1.11.0a8](https://github.com/astronomer/astronomer-cosmos/releases/tag/astronomer-cosmos-v1.11.0a8), we identified that it was not working as expected.

There were two main issues:
1. We were not setting the `producer_task_id` argument correctly in the sensor tasks
2. The producer task was always being picked up after the sensor tasks, even though it had a higher priority weight and was an "upstream" task of the dbt root nodes (with a trigger rule always, so they could start sensing).

The first issue was a bug, which is now fixed.

The second issue occurs because the sensor tasks have a `trigger_rule` of "always". This setting works well in the case of `DbtDag`, as the only upstream task is the producer task, and the producer task and root dbt project nodes can be scheduled at the same time, with the task weight of the producer giving its precedence. However, in the case of a `DbtTaskGroup` with one or multiple upstream tasks, the producer task must wait for the upstream task, whereas the dbt project root nodes ("top sensors") do not have to wait.  This leads to sensors occupying the available lots before the producer is scheduled to arrive.

At some point, we suspected this was an issue with Airflow itself, but after a troubleshooting session with Ash, we realised it was not a bug in Airflow: https://github.com/apache/airflow/issues/56723.

The side-effect of the implementation in this PR is that `DbtTaskGroup` and `DbtDag` are behaving differently:
- In `DbtDag`, the producer is upstream of the dbt root nodes
- In `DbtTaskGroup`, the producer is not an upstream root node

The reason we are keeping this behaviour in `DbtDag` is that it is not leading to any issues. It is topologically "correct", making it easier for users to see the dependency between the tasks & potentially view the task list topologically sorted.

At this moment, we were unable to find a way to offer this same behaviour with `DbtTaskGroup`. @ashb , with whom I troubleshooted the original "problem", pointed out Airflow could potentially have a different "state" that wouldn't always be, but something like "start when my upstream task started".  However, by examining [Airflow's current implementation](https://github.com/apache/airflow/blob/c160ab70a00faa41ef5996c4ba4a6580a64194ad/airflow-core/src/airflow/ti_deps/deps/trigger_rule_dep.py), any changes to this area could potentially incur an additional cost to the scheduler and compromise Airflow's performance.

If we were able to "force" in the Airflow UI to have the producer task on the "top" both in the graph view and in the list of functions view, we could make `DbtDag`and `DbtTaskGroup` more consistent (by not having the producer task as upstream and dbt root nodes with trigger_rule always). Currently, in Airflow 3.0.x and Airflow 3.1.0, the producer task has been showing randomly, unfortunately, but we will work with the Airflow team to improve this and add follow-up tickets as needed.

Closes: https://github.com/astronomer/astronomer-cosmos/issues/1961
Closes: https://github.com/astronomer/oss-integrations-private/issues/240

tatiana committed:
Fix running ExecutionMode.WATCHER in DbtTaskGroup

tatiana committed:
Fix tests and DbtDag WATCHER dependency

tatiana committed:
Add test for watcher with TaskGroup

tatiana committed:
Comment out example DAG related to watcher in TaskGroup

tatiana committed:
Fix test_dbt_task_group_with_watcher in AF 2.9

tatiana committed:
Remove example_watcher_taskgroup since it doesn't work with dag.test()

tatiana committed:
Merge branch 'main' into fix-watcher-taskgroup

tatiana committed:
Apply suggestion from @tatiana

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

tatiana committed:
Apply suggestion from @tatiana

tatiana committed:
Apply suggestion from @tatiana

tatiana commented on cosmos/airflow/graph.py:
```diff
@@ -567,11 +567,28 @@ def _add_producer_watcher(
+        )
+
+        # the following only works with DbtDag. It does not work with DbtTaskGroup due to an Airflow bug
+        if "DbtDag" in dag.__class__.__name__:
```
We cannot use `isinstance` because it would lead to a circular dependency.

tatiana committed:
Remove unintended commited dags folder

tatiana committed:
Fix task dependencies

pankajkoti commented on cosmos/airflow/graph.py:
```diff
@@ -567,11 +567,28 @@ def _add_producer_watcher(
+
+    # Consumer tasks will need to be updated to use the producer task as a dependency
+    for node_id, task_or_taskgroup in tasks_map.items():
         # we want to make the producer task to be the parent of the root dbt nodes, without blocking them from sensing XCom
```
Can we ensure these comments are valid for `DbtTaskGroup` too?

pankajkoti reviewed (approved):
Happy to merge if we've validated the behaviour. Don't have time myself today to validate the issue and the fix.

tatiana commented on cosmos/airflow/graph.py:
```diff
@@ -567,11 +567,28 @@ def _add_producer_watcher(
+
+        # We also need to set the producer task ID to all consumer tasks, regardless of whether they are root or not
+        for task in node_tasks:
+            task.producer_task_id = producer_airflow_task.task_id  # type: ignore[attr-defined]
```
Yes! Previously the producer taks was being instantiated before, and the logic wouldn't work

tatiana committed:
Fix Cosmos behaviour when users set invocation_mode=InvocationMode.DBT_RUNNER

```
With 1.11.0a8, Cosmos would fail with:
  File "/Users/tatiana.alchueyr/Code/astronomer-cosmos/dags/example_watcher.py", line 46, in <module>
    execution_config=ExecutionConfig(
                     ^^^^^^^^^^^^^^^^
  File "<string>", line 9, in __init__
  File "/Users/tatiana.alchueyr/Code/astronomer-cosmos/cosmos/config.py", line 438, in __post_init__
    raise CosmosValueError(
cosmos.exceptions.CosmosValueError: ExecutionConfig.invocation_mode is only configurable for ExecutionMode.LOCAL and ExecutionMode.VIRTUALENV.
```

If users attempted to run the DbtDag or TaskGroup setting the invocation_mode=InvocationMode.DBT_RUNNER:

```
from cosmos.constants import InvocationMode

example_watcher = DbtDag(
    # dbt/cosmos-specific parameters
    execution_config=ExecutionConfig(
        execution_mode=ExecutionMode.WATCHER,
        invocation_mode=InvocationMode.DBT_RUNNER
    ),
    project_config=ProjectConfig(DBT_PROJECT_PATH),
    profile_config=profile_config,
    operator_args=operator_args,
    # normal dag parameters
    schedule="@daily",
    start_date=datetime(2023, 1, 1),
    catchup=False,
    dag_id="example_watcher",
    default_args={"retries": 0},
)
```

This PR fixes this.

tatiana committed:
Improve test coverage

tatiana committed:
Merge branch 'main' into fix-watcher-taskgroup

tatiana committed:
Fix behaviour of which nodes should have trigger_rule always

tatiana committed:
Make CI happy with tests
